resharper/resharper-unity/src/Rider/UnitTesting/UnityTaskRunnerHostControllerExtension.cs

```diff
@@ -0,0 +1,130 @@
+using JetBrains.Util;
+
+namespace JetBrains.ReSharper.Plugins.Unity.Rider.UnitTesting
+{
+    [SolutionComponent]
+    public class UnityTaskRunnerHostControllerExtension : ITaskRunnerHostControllerExtension
+    {
+        private static readonly Key<LifetimeDefinition> ourLifetimeDefinitionKey = new Key<LifetimeDefinition>("UnityTaskRunnerHostController.CancelPrepareForRun");
+        
+        private const string PluginName = "Unity plugin";
+        private const string NotAvailableUnityEditorMessage = "Unable to {0} tests: Unity Editor is not running";
+        private const string StartUnityEditorQuestionMessage = "To {0} unit tests, you should first run Unity Editor. Do you want to Start Unity {1} now?";
+
+        private readonly Lifetime myLifetime;
+        private readonly IUnityController myUnityController;
+        private readonly IBackgroundProgressIndicatorManager myBackgroundProgressIndicatorManager;
+        private readonly IShellLocks myShellLocks;
+        private readonly IDictionary<string, string> myAvailableProviders;
+        private readonly object myStartUnitySync = new object();
+        private Task myStartUnityTask;
+
+        public UnityTaskRunnerHostControllerExtension(Lifetime lifetime,
+                                                      IShellLocks shellLocks,
+                                                      IUnityController unityController,
+                                                      IBackgroundProgressIndicatorManager backgroundProgressIndicatorManager)
+        {
+            myLifetime = lifetime;
+            myShellLocks = shellLocks.NotNull();
+            myUnityController = unityController.NotNull();
+            myBackgroundProgressIndicatorManager = backgroundProgressIndicatorManager.NotNull();
+            myStartUnityTask = Task.CompletedTask;
+            myAvailableProviders = new Dictionary<string, string>
+            { 
+                { WellKnownHostProvidersIds.RunProviderId, "Run" },
+                { WellKnownHostProvidersIds.DebugProviderId, "Debug" }
+            };
+        }
+
+        public bool IsApplicable(IUnitTestRun run)
+        {
+            var isUnity = myUnityController.IsUnityEditorUnitTestRunStrategy(run.RunStrategy);
+            return isUnity && myAvailableProviders.ContainsKey(run.HostController.HostId);
+        }
+
+        public ClientControllerInfo GetClientControllerInfo(IUnitTestRun run, ITaskRunnerHostController next) => null;
+
+        public Task PrepareForRun(IUnitTestRun run, ITaskRunnerHostController next)
+        {
+            var lifetimeDef = myLifetime.CreateNested();
+            run.PutData(ourLifetimeDefinitionKey, lifetimeDef);
+
+            lock (myStartUnitySync)
+            {
+                WrapStartUnityTask(() => PrepareForRunInternal(lifetimeDef.Lifetime, run));
+                WrapStartUnityTask(() => next.PrepareForRun(run));
+                
+                return myStartUnityTask;
```

Comment:
I don't understand all of this task wrapping. Couldn't it just be?
```
await PrepareForRunInternal(lifetimeDef.Lifetime, run);
await next.PrepareForRun(run);
```
Or did you explicitly want to run both task (yourself and next) concurrently?
``` 
return Task.WhenAll(
  PrepareForRunInternal(lifetimeDef.Lifetime, run),
  next.PrepareForRun(run)
);
```
Also, why create nested lifetime inside this method, if it is only used it PrepareForRunInternal?
It seems unnecessary complicated